src/Arduino_NotecardConnectionHandler.cpp

```diff
@@ -0,0 +1,979 @@
+      transport_connected  = (x & 0x01);
+      connected_to_notehub = (x & 0x02);
+      notecard_error       = (x & 0x04);
+      host_error           = (x & 0x08);
+      reserved             = (x & 0xF0);
+      return *this;
+  }
+  operator uint_fast8_t () const {
+      return ((reserved << 4) | (host_error << 3) | (notecard_error << 2) | (connected_to_notehub << 1) | (transport_connected));
+  }
+
+  bool transport_connected  : 1;
+  bool connected_to_notehub : 1;
+  bool notecard_error       : 1;
+  bool host_error           : 1;
+  uint_fast8_t reserved     : 4;
+};
+static_assert(sizeof(NotecardConnectionStatus) == sizeof(uint_fast8_t));
+
+/******************************************************************************
+   CTOR/DTOR
+ ******************************************************************************/
+
+NotecardConnectionHandler::NotecardConnectionHandler(
+  const String & project_uid_,
+  bool en_hw_int_,
+  bool keep_alive_,
+  uint32_t i2c_address_,
+  uint32_t i2c_max_,
+  TwoWire & wire_,
+  const String & notehub_url_
+) :
+  ConnectionHandler{keep_alive_, NetworkAdapter::NOTECARD},
+  _serial(nullptr),
+  _wire(&wire_),
+  _inbound_buffer(nullptr),
+  _conn_start_ms(0),
+  _i2c_address(i2c_address_),
+  _i2c_max(i2c_max_),
+  _uart_speed(0),
+  _inbound_buffer_index(0),
+  _inbound_buffer_size(0),
+  _en_hw_int(en_hw_int_),
+  _topic_type{TopicType::Invalid},
+  _notecard{},
+  _device_id{},
+  _notecard_uid{},
+  _notehub_url(notehub_url_),
+  _project_uid(project_uid_)
+{ }
+
+NotecardConnectionHandler::NotecardConnectionHandler(
+  const String & project_uid_,
+  HardwareSerial & serial_,
+  uint32_t speed_,
+  bool en_hw_int_,
+  bool keep_alive_,
+  const String & notehub_url_
+) :
+  ConnectionHandler{keep_alive_, NetworkAdapter::NOTECARD},
+  _serial(&serial_),
+  _wire(nullptr),
+  _inbound_buffer(nullptr),
+  _conn_start_ms(0),
+  _i2c_address(0),
+  _i2c_max(0),
+  _uart_speed(speed_),
+  _inbound_buffer_index(0),
+  _inbound_buffer_size(0),
+  _en_hw_int(en_hw_int_),
+  _topic_type{TopicType::Invalid},
+  _notecard{},
+  _device_id{},
+  _notecard_uid{},
+  _notehub_url(notehub_url_),
+  _project_uid(project_uid_)
+{ }
+
+/******************************************************************************
+   PUBLIC MEMBER FUNCTIONS
+ ******************************************************************************/
+
+const String & NotecardConnectionHandler::syncArduinoDeviceId (const String & device_id_)
+{
+  // Validate the connection state is not in an initialization state
+  if (check() == NetworkConnectionState::INIT)
+  {
+    Debug.print(DBG_ERROR, F("Failed to sync Arduino Device ID. Connection has not been initialized."));
+    return device_id_;
+  }
+
+  // Prefer the incoming Device ID. If the value is known to be invalid, then
+  // fetch the Device ID from the Notecard Connection Handler (cloud cache)
+  if ((device_id_ == "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" || device_id_.length() == 0)
```

Comment:
As discussed i think it would be better to move the `device_id_`  and `secret_device_key_` inside the ArduinoCloud Notecard implementation. The connectionHandler should not be aware about device authentication. Here is where we get the `device_id` and `certificate` for the TCP implementation. https://github.com/arduino-libraries/ArduinoIoTCloud/blob/6ef7ad86afee69987af8ef57fbad5c1e2224edde/src/ArduinoIoTCloudTCP.cpp#L127

Reply:
The introduction of external connection management breaks the standard paradigm.

Attempting to map the existing organization to the constraints of the hardware is creating a dependency problem in the code.

Ultimately, this may still be necessary, but perhaps we can discuss more.

Reply:
We have removed the concept of device id and secret management from the host device, so this code has been removed completely.